solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/CollectionPropertiesZkStateReader.java

```diff
@@ -0,0 +1,403 @@
+   * @param cacheForMillis The minimum number of milliseconds to maintain a cache for the specified
+   *     collection's properties. Setting a {@code CollectionPropsWatcher} will override this value
+   *     and retain the cache for the life of the watcher. A lack of changes in zookeeper may allow
+   *     the caching to remain for a greater duration up to the cycle time of {@code CacheCleaner}.
+   *     Passing zero for this value will explicitly remove the cached copy if and only if it is due
+   *     to expire and no watch exists. Any positive value will extend the expiration time if
+   *     required.
+   * @return a map representing the key/value properties for the collection.
+   */
+  public Map<String, String> getCollectionProperties(final String collection, long cacheForMillis) {
+    synchronized (watchedCollectionProps) { // synchronized on the specific collection
+      Watcher watcher = null;
+      if (cacheForMillis > 0) {
+        watcher =
+            collectionPropsWatchers.compute(
+                collection,
+                (c, w) ->
+                    w == null ? new PropsWatcher(c, cacheForMillis) : w.renew(cacheForMillis));
+      }
+      VersionedCollectionProps vprops = watchedCollectionProps.get(collection);
+      boolean haveUnexpiredProps = vprops != null && vprops.cacheUntilNs > System.nanoTime();
+      long untilNs =
+          System.nanoTime() + TimeUnit.NANOSECONDS.convert(cacheForMillis, TimeUnit.MILLISECONDS);
+      Map<String, String> properties;
+      if (haveUnexpiredProps) {
+        properties = vprops.props;
+        vprops.cacheUntilNs = Math.max(vprops.cacheUntilNs, untilNs);
+      } else {
+        try {
+          VersionedCollectionProps vcp = fetchCollectionProperties(collection, watcher);
+          properties = vcp.props;
+          if (cacheForMillis > 0) {
+            vcp.cacheUntilNs = untilNs;
+            watchedCollectionProps.put(collection, vcp);
+          } else {
+            // we're synchronized on watchedCollectionProps and we can only get here if we have
+            // found an expired vprops above, so it is safe to remove the cached value and let the
+            // GC free up some mem a bit sooner.
+            if (!collectionPropsObservers.containsKey(collection)) {
+              watchedCollectionProps.remove(collection);
+            }
+          }
+        } catch (Exception e) {
+          throw new SolrException(
+              SolrException.ErrorCode.SERVER_ERROR,
+              "Error reading collection properties",
+              SolrZkClient.checkInterrupted(e));
+        }
+      }
+      return properties;
+    }
+  }
+
+  @Override
+  public void close() {
+    this.closed = true;
+    notifications.shutdownNow();
+    ExecutorUtil.shutdownAndAwaitTermination(notifications);
+    ExecutorUtil.shutdownAndAwaitTermination(collectionPropsNotifications);
+
+    assert ObjectReleaseTracker.release(this);
```

Comment:
I don't recommend bothering with this; it's only closely associated with ZkStateReader so the accidental risk is extremely low.

Reply:
Removed